{% extends "layouts/layout.html" %}
{%- from 'macros/nav.html' import nav -%}

{% block body %}
{{ nav("home") }}
<main>
    <section>
        <h2>Welcome to CycleDash <img class="cycle" src="static/img/cycle-dash.png" width=100></h2>
        <p>
            Cycledash is an open-source somatic variant caller analysis and
            visualization tool. The
            <a href="https://github.com/hammerlab/cycledash">source can be found</a>
            on GitHub, and is developed by the
            <a href="http://hammerlab.org">Hammerlab</a> out of Mount Sinai's Icahn
            School of Medicine.
        </p>
        <p>Runs can be found at <tt><a href="/runs">/runs</a></tt>.</p>
    </section>
    <section>
        <h3>JSON API</h3>
        <p>The primary endpoint for posting data to from an external source is /runs.</p>
        <p>JSON should be POSTed to /runs with following fields:</p>

        <h4>Required</h4>
        <dl class="dl-horizontal">
            <dt>vcfPath</dt> <dd>The path on HDFS where the VCF can be found. This should be immutable, as CycleDash expects to be able to find the VCF here at any time.</dd>
            <dt>variantCallerName</dt> <dd>The name of the variant caller which produced this VCF. This should remain constrant between VCFs with the same caller in order to compare runs to one another.</dd>
        </dl>

        <h4>Optional</h4>
        <dl class="dl-horizontal">
            <dt>truthVcfPath</dt> <dd>The path on HDFS for the truth (or "reference") VCF. This should be immutable.</dd>
            <dt>dataset</dt> <dd>The name of the dataset on which the caller was run (e.g. Dream Chromosome 20).</dd>
            <dt>tumorPath</dt> <dd>The path on HDFS of the tumor BAM on which the caller was run.</dd>
            <dt>normalPath</dt> <dd>The path on HDFS of the normalBAM on which the caller was run.</dd>
            <dt>params</dt> <dd>Params that the caller was run with, or other notes relevant to the run.</dd>
        </dl>
    </section>
</main>
{% endblock %}
